### ISSUE TITLE
Hidden controls SC3.2.7 - Does this cover invisible form fields?

### ISSUE THREAD

**Issue by guyhickling** [w3c/wcag#1243]
A design pattern that is becoming increasingly popular these days among the more sadistic and antisocial web designers is where they create input forms showing just the field labels, usually as a vertical list, but seemingly no fields. It is assumed the user will work out that there are invisible fields to the right of the labels. But you don't know that, until you hover over or tab to the blank page beside each label and notice that the cursor changes to a vertical bar indicating you are now over an input field. I have audited about half a dozen websites like this in recent times. Sometimes, if the label texts don't obviously imply fields, less alert users can miss the entire form!

Does that design pattern fail this SC? My reading of the Understanding doc tells me that it does. So there should be indicators for each field present at the start (or a button or something to make all the fields visible on demand).

If it fails, could that fact be explicitly mentioned in the Understanding doc in the list of Failure techniques (I'll be happy to draft a Technique document for it), so that testers know whether they can or cannot fail it. And so developers can't push back, because this weird design does seem to be becoming more common.

**Comment by awkawk** [comment]
This sounds more like Non-text contrast if you can't tell where the control is.

**Comment by guyhickling** [comment]
No, when I say invisible, I mean that literally - there is absolutely no marker for the field at all. No border, no underline, no nothing. So the Non-text Contrast SC doesn't apply. Only the text label or labels suggest there might be a field there. Yes, bonkers of course, but it seems to appeal to those minimalist designers who want to do away with everything that makes a web page a web page and confuse their users as much as they possibly can!

**Comment by JAWS-test** [comment]
It would be nice to have a criterion that prohibits invisible form fields. However, SC 3.2.7 does not seem to me to be suitable for this in its current form, as it contains the wording "Controls needed to progress or complete a process"

Since not every form field is necessary to complete a process (since not every form field is a mandatory field), there would be a gap. 

SC 1.4.11 "Non-text Contrast" does not seem to me to solve the problem either, since it only applies if anything is visible at all

**Comment by mraccess77** [comment]
I run into this issue frequently - we tried to get an affordance SC on this topic but were not able to get that in for WCAG 2.2.

**Comment by guyhickling** [comment]
The way I look at it is that completing any form _is_ a process. It achieves some purpose of passing on data. And the fields on a form are all part of that process. Even (looking at the glossary definition of process) if some or all of them are not required because some users may not want to enter data for those fields, they are still a necessary part of that process, needed to get that data when it is applicable but not when that data isn't applicable or a default will be used. But I understand that it might be argued even so. 

@mraccess77 The simplest thing would be to add three words "or input forms" to this Hidden Controls SC to make:
> Controls needed to progress or complete a process or input form

-do you think we could get that agreed by the WG, given that the SC is being worked on right now?

**Comment by patrickhlauke** [comment]
if you squint and look hard enough, *everything* can become a "process" ...

**Comment by alastc** [comment]
(Chair hat off) I'd say that controls does cover inputs, but permanently hidden controls are not covered by this SC because they are not shown with "pointer hover or keyboard focus".

**Comment by guyhickling** [comment]
Ah but that's the point - they aren't permanently hidden, they are actually there, just not visible - they need hover or focus to reveal them - which is the point of this SC. You see the column of label texts on the page and the designer expects users to hover over the area to discover the fields. You hover and the cursor changes indicating something is there. 

You then click on it and the input cursor changes to the one normally expected in input fields (blinking vertical bar). If you are really lucky and have met up with a designer that shows a modicum of remorse, an underline will appear to show the length of the field's typing area! You type and find out it lets you do that. Yes, a crackpot design but hey, I didn't invent the 'orrible thing, I've just met it a few times!

But since hovering or focus reveals the presence of the field, they aren't actually hidden. So I think on that basis it comes under the SC, and therefore needs a mechanism to be offered to the user to make the fields persistently visible.

**Comment by alastc** [comment]
> they need hover or focus to reveal them... You hover and the cursor changes indicating something is there.

So to refine my answer, I don't think a cursor change is relevent here. (Some interfaces don't have a cursor.)

If an underline appears on hover/focus, then it would fail. If nothing appears (of the element) then it doesn't fail the SC now (in my reading).

Have you come across any that literally show nothing, even on focus/hover?

**Comment by mraccess77** [comment]
The common example is a name and when you hover or focus the name, an edit link appears or the name label morphs into a name edit field with a box around it..

**Comment by guyhickling** [comment]
@mraccess77 Yes, there's that pattern as well. That is another example of the Gmail type of situation raised in #1241. In the Gmail inbox you don't hover over the 4 buttons (because they aren't present at that point, they are hidden using display:none), you hover over any part the text showing for one of the emails in inbox list, then the 4 buttons get overlaid on top of the last inch or so of that text. 

As the SC is worded at the moment, this pattern you mention in forms, and the Gmail inbox which involves JavaScript button controls, would both not be caught by this SC due to its wording. This whole SC needs to have a few extra words added to make it really useful to disabled people by making it more generic and inclusive of different types of pattern. At present it is far too restrictive, catching only a few of the possible variations that we see daily out in the field.

@alastc How about it?

**Comment by StommePoes** [comment]
Does the [pattern used by Gutenberg](https://wordpress.org/gutenberg/) match this use case or no? In this case it's not the controls themselves being hovered/focussed, but an editable region. I'm thinking no, but...

Looking at Web Editor Success: "In a web editor, once a page is open for editing, all the possible locations for editing display a pencil icon."

I have indeed seen the case described in Guy's first post, and with screen magnification and using a mouse, this lack of visible hit areas is really disconcerting.
I would get this under the old Gmail inputs with WHC (which I think wouldn't fail any new SCs) because they used a div with a CSS background colour to denote a fake "bottom border". So the effect as a user was the same as Guy's case above with no visible hit area until focus/hover. These require either randomly waving a mouse around or sorta forcing you to use the Tab key so you can see the inputs.

**Comment by detlevhfischer** [comment]
Proposed working group answer:
Thank you for your comment.
The definition of process ("series of user actions where each action is required in order to complete an activity") implies that at a minimum, all required fields must be visible by default without keyboard or pointer focus. The addition to the SC text that you suggest ("Controls needed to progress or complete a process **or input form**") seems unnecessary. The generic term 'process' has been chosen to encompass all elements that may be needed by the user to proceed, whatever the exact process. In a form-based process, these elements include the input fields. While a narrow reading of the SC text may suggest that fields that are not required can be hidden, we think that there will be little incentive to do so if required fields must be shown in any case.

**Comment by guyhickling** [comment]
I understand what you say about not needing to alter the SC text itself. So could we add something to the Understanding doc to make it clear that invisible forms are covered by this SC? Otherwise I'm worried that too many developers (and a11y consultants as well) will try to claim the new design pattern mentioned in the original post above is ok. They will try to argue fields aren't really "controls", or they aren't part of a process, or something like that. 

I suggest the easiest way to cover it is to add an example covering input forms in the Examples section as the least intrusive way of adding to the doc. How about:

> Input Form Success: In an input form, all the text fields have a border round them to indicate where typing of data is required.

That doesn't prevent designers who want to show other indicators (such as underlines or overlines rather than field borders which is also a growing trend), but it offers the best kind of design as the example, and will confirm that fields that are invisible until hover are a fail of this SC.

**Comment by patrickhlauke** [comment]
form fields have been called "controls" for aeons though, even directly in the HTML spec (think 4.0 was the earliest that called them that https://www.w3.org/TR/html401/interact/forms.html#h-17.2). not sure it needs specifying/clarifying, as that's always been the meaning of "control", and other parts of WCAG use it with this meaning already as well (see for instance the whole of https://www.w3.org/TR/WCAG21/#input-purposes that talks about controls when really it's referring to inputs/textareas/selects that can take `autocomplete` attributes)

**Comment by StommePoes** [comment]
I'd still like the Understanding to have a form-control example. Makes it clear.

**Comment by alastc** [comment]
Hi Guy,

The [latest version](https://raw.githack.com/w3c/wcag/wcag22-visible-controls-update/understanding/22/visible-controls.html) uses a different approach which focuses on the information for that control being available is visible.

That would mean the case of invisible inputs is covered.

Please do raise any new issues on the new text, it will be going out for review again.
